source/language/openpulse.rst

```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
-   gaussian_square(length:l, complex[float[32]]:amp, length:center, length:square_width, length:sigma)
-
-   // amp is pulse amplitude at center
-   // center is the mean of pulse
-   // sigma is the standard deviation of pulse
-   // beta is the Y correction amplitude, see the DRAG paper
-   drag(length:l, complex[float[32]]:amp, length:center, length:sigma, float[32]:beta)
+Waveforms are of type ``waveform`` and can either be:
+
+- An array of complex samples which define the points for the waveform envelope
+- An abstract mathematical function representing a waveform. This will later be
+  materialized into a list of complex samples, either by the compiler or the hardware
+  using the parameters provided to the pulse template.
+
+A value of type ``waveform`` is retrieved by explicitly constructing the complex samples
+or by calling one of the built-in waveform template functions. Note that each of these
+functions takes a type ``length`` as a first argument, since waveforms need to have a definite
+length. Using the hardware dependent ``dt`` unit is recommended, since the compiler may need to
+down-sample a higher precision waveform to physically realize it.
 
-   // Define a constant pulse of length l
-   constant(l:length)
+.. code-block:: javascript
+
+   // arbitrary complex samples
+   arb_waveform = [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j];
+
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // sigma is the standard deviation of waveform
+   gaussian(complex[float[32]]:amp, length:l, length:sigma)
+
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // sigma is the standard deviation of waveform
+   sech(complex[float[32]]:amp, length:l, length:sigma)
+
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // square_width is the width of the square waveform component
+   // sigma is the standard deviation of waveform
+   gaussian_square(complex[float[32]]:amp, length:l, length:square_width, length:sigma)
+
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // sigma is the standard deviation of waveform
+   // beta is the Y correction amplitude, see the DRAG paper
+   drag(complex[float[32]]:amp, length:l, length:sigma, float[32]:beta)
+
+   // amp is waveform amplitude
+   // l is the overall length of the waveform
+   constant(complex[float[32]]:amp, length:l)
+
+   // amp is waveform amplitude
+   // l is the overall length of the waveform
+   // frequency is the frequency of the waveform
+   // phase is the phase of the waveform
+   sine(complex[float[32]]:amp, length: l, float[32]: frequency, angle: phase)
+
+We can manipulate the ``waveform`` types using the following signal processing functions to produce
+new waveforms
+
+- ``mix(wf1: waveform, wf2: waveform)`` -> waveform - Mix two input waveforms to produce a new waveform.
```

Comment:
mix and scale are the only uses I see in the examples, makes me wonder whether it's worth postponing adding some of these other ones

Reply:
Because frames are cheap now we may not need it but I can see folks wanting `sum` and `piecewise`.

Reply:
`phase` may be useful to produce a Y from an X pulse, but this could be equivalently realized with a frame.